### PR TITLE
Allow region-ized fields in OpenStack CloudProfileConfig

### PR DESCRIPTION
**Pull request by rfranzke**
**What this PR does / why we need it**:
Allow `region`-ized fields in OpenStack `CloudProfileConfig`. Some OpenStack environments have dedicated keystone URLs, floating pools or load balancer per region/data center. Today, you would have to create a dedicated `CloudProfile` per region (although the backing cloud provider is the same). That's unfortunate. With this PR, it's possible to configure regions for the mentioned properties. This will allow to have only one `CloudProfile` like for the hyperscalers.

**Special notes for your reviewer**:
* /cc @afritzler 
* On purpose, I didn't change the type of `KeystoneURL` from `string` to `*string` (because it's optional now) to prevent vendoring issues. We could do that later if we want.
```
# github.com/gardener/gardener/pkg/apis/core/v1alpha1
vendor/github.com/gardener/gardener/pkg/apis/core/v1alpha1/conversions.go:580:34: cannot use cloudProfileConfig.KeyStoneURL (type *string) as type string in assignment
# github.com/gardener/gardener/pkg/apis/core/v1beta1
vendor/github.com/gardener/gardener/pkg/apis/core/v1beta1/conversions.go:507:34: cannot use cloudProfileConfig.KeyStoneURL (type *string) as type string in assignment
make: *** [start-provider-openstack] Error 2
```

**Release note**:

```noteworthy operator
The OpenStack `CloudProfileConfig` now features a new `keyStoneURLs` field. It is a mapping of regions to keystone URLs. Also, the `constraints.loadBalancerProviders` and `constraints.floatingPools` fields now can optionally have a `region` property. This allows to only have one `CloudProfile` for an OpenStack environment in which the keystone URL, floating pools, and/or load balancer providers differ per region. See [this document](https://github.com/gardener/gardener-extensions/blob/master/controllers/provider-openstack/docs/usage-as-operator.md#cloudprofileconfig) for an example resource. :important: Please note that these fields will only be usable with gardener/gardener once the deprecated `garden.sapcloud.io/v1beta1` API group has been removed.
```

### CODE REVIEW

**Review by afritzler — COMMENTED** [review]
Looks good to me, but I wanted to do some testing next week.

**Review by timuthy — COMMENTED** [review]
Nice PR! I have a general comment about the setting override, though.
For the `keyStoneURL` the non region specific setting overrides the region specifics and for `LoadBalancerProvider` and `FloatingPool` the order of elements determines how it is overwritten. I think this should be a) consistently handled and b) the region specifics should overrule the non-region specifics. WDYT?

**Comment by rfranzke** [comment]
As discussed with @timuthy I will adapt the PR so that it's allowed to specify both the non-regional and regional fields (e.g., `keystoneURL` and `keystoneURLs`). The logic will always prefer a regional setting if found, if not it will fallback to the non-regional setting.

File: controllers/provider-openstack/pkg/apis/openstack/validation/cloudprofile.go [hunk 1]
```diff
@@ -74,10 +101,28 @@ func ValidateCloudProfileConfig(cloudProfile *apisopenstack.CloudProfileConfig)
 		}
 	}
 
-	if len(cloudProfile.KeyStoneURL) == 0 {
+	if len(cloudProfile.KeyStoneURL) == 0 && len(cloudProfile.KeyStoneURLs) == 0 {
```
[the remark this answers is not in the document]
Comment (rfranzke) [reply controllers/provider-openstack/pkg/apis/openstack/validation/cloudprofile.go]: As discussed: no, let's allow specifying both to be able to use `keystoneURL` as fallback.

**Comment by ialidzhikov** [comment]
Looks good to me.
Nit - rebase needed because of the new year.

**Comment by rfranzke** [comment]
Thanks, merging after rebase.

**Pull request merged**